File: internal/updater/discovery_test.go
package updater

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync/atomic"
	"testing"
	"time"
)

func TestCheckSelectsExactPackagedAsset(t *testing.T) {
	t.Parallel()
	tests := []struct {
		name   string
		target Target
		asset  string
	}{
		{name: "mac Intel", target: Target{GOOS: "darwin", GOARCH: "amd64"}, asset: "p-track_1.2.4_darwin_amd64.dmg"},
		{name: "mac Apple silicon", target: Target{GOOS: "darwin", GOARCH: "arm64"}, asset: "p-track_1.2.4_darwin_arm64.dmg"},
		{name: "Linux Intel", target: Target{GOOS: "linux", GOARCH: "amd64"}, asset: "ptrack_1.2.4_linux_amd64.tar.gz"},
		{name: "Linux ARM", target: Target{GOOS: "linux", GOARCH: "arm64"}, asset: "ptrack_1.2.4_linux_arm64.tar.gz"},
		{name: "Windows Intel", target: Target{GOOS: "windows", GOARCH: "amd64"}, asset: "ptrack_1.2.4_windows_amd64.zip"},
		{name: "Windows ARM", target: Target{GOOS: "windows", GOARCH: "arm64"}, asset: "ptrack_1.2.4_windows_arm64.zip"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()
			client := fixtureClient(t, releaseJSON("v1.2.4", false, false, tt.asset))
			candidate, err := client.Check(context.Background(), "v1.2.3", tt.target)
			if err != nil {
				t.Fatal(err)
			}
			if candidate.Version != "1.2.4" || candidate.Tag != "v1.2.4" || candidate.Package.Name != tt.asset {
				t.Fatalf("candidate = %#v", candidate)
			}
			if candidate.Checksums.Name != "checksums.txt" || candidate.PageURL != "https://github.com/ro-ag/ptrack/releases/tag/v1.2.4" {
				t.Fatalf("candidate metadata = %#v", candidate)
			}
		})
	}
}

func TestCheckRejectsDevelopmentUnsupportedAndNonNewerVersions(t *testing.T) {
	t.Parallel()
	client := fixtureClient(t, releaseJSON("v1.2.4", false, false, "ptrack_1.2.4_linux_arm64.tar.gz"))
	for _, current := range []string{"dev", "", "1.2", "1.2.3-beta.1", "01.2.3", "vv1.2.3"} {
		if _, err := client.Check(context.Background(), current, Target{GOOS: "linux", GOARCH: "arm64"}); !errors.Is(err, ErrDevelopmentBuild) {
			t.Errorf("current %q: error = %v, want ErrDevelopmentBuild", current, err)
		}
	}
	for _, current := range []string{"1.2.4", "1.2.5", "2.0.0"} {
		if _, err := client.Check(context.Background(), current, Target{GOOS: "linux", GOARCH: "arm64"}); !errors.Is(err, ErrNoUpdate) {
			t.Errorf("current %q: error = %v, want ErrNoUpdate", current, err)
		}
	}
	for _, target := range []Target{{GOOS: "freebsd", GOARCH: "amd64"}, {GOOS: "linux", GOARCH: "386"}, {GOOS: "", GOARCH: ""}} {
		if _, err := client.Check(context.Background(), "1.2.3", target); !errors.Is(err, ErrUnsupportedTarget) {
			t.Errorf("target %#v: error = %v, want ErrUnsupportedTarget", target, err)
		}
	}
}

func TestCheckRejectsUnsafeReleaseMetadata(t *testing.T) {
	t.Parallel()
	asset := "ptrack_1.2.4_linux_amd64.tar.gz"
	valid := releaseJSON("v1.2.4", false, false, asset)
	tests := []struct {
		name string
		body string
	}{
		{name: "draft", body: releaseJSON("v1.2.4", true, false, asset)},
		{name: "prerelease flag", body: releaseJSON("v1.2.4", false, true, asset)},
		{name: "prerelease tag", body: releaseJSON("v1.2.4-rc.1", false, false, asset)},
		{name: "missing package", body: strings.ReplaceAll(valid, asset, "ptrack_1.2.4_linux_arm64.tar.gz")},
		{name: "duplicate package", body: duplicateAsset(valid, asset)},
		{name: "duplicate checksum", body: duplicateAsset(valid, "checksums.txt")},
		{name: "wrong package host", body: strings.Replace(valid, "https://github.com/ro-ag/ptrack/releases/download/v1.2.4/"+asset, "https://evil.example/"+asset, 1)},
		{name: "wrong package path", body: strings.Replace(valid, "/ro-ag/ptrack/releases/download/", "/ro-ag/other/releases/download/", 1)},
		{name: "package query", body: strings.Replace(valid, asset+`","size"`, asset+`?token=secret","size"`, 1)},
		{name: "empty package query", body: strings.Replace(valid, asset+`","size"`, asset+`?","size"`, 1)},
		{name: "package fragment", body: strings.Replace(valid, asset+`","size"`, asset+`#fragment","size"`, 1)},
		{name: "empty package fragment", body: strings.Replace(valid, asset+`","size"`, asset+`#","size"`, 1)},
		{name: "package port", body: strings.Replace(valid, "https://github.com/", "https://github.com:443/", 1)},
		{name: "package userinfo", body: strings.Replace(valid, "https://github.com/", "https://user@github.com/", 1)},
		{name: "wrong checksum path", body: strings.Replace(valid, "/checksums.txt", "/other.txt", 1)},
		{name: "pending asset", body: strings.Replace(valid, `"state":"uploaded"`, `"state":"new"`, 1)},
		{name: "empty asset", body: strings.Replace(valid, `"size":1024`, `"size":0`, 1)},
		{name: "oversized package", body: strings.Replace(valid, `"size":1024`, fmt.Sprintf(`"size":%d`, maxAssetBytes+1), 1)},
		{name: "oversized checksum", body: strings.Replace(valid, `"size":128`, fmt.Sprintf(`"size":%d`, maxManifestBytes+1), 1)},
		{name: "missing published time", body: strings.Replace(valid, `"published_at":"2026-08-11T00:00:00Z"`, `"published_at":"0001-01-01T00:00:00Z"`, 1)},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()
			client := fixtureClient(t, tt.body)
			if _, err := client.Check(context.Background(), "1.2.3", Target{GOOS: "linux", GOARCH: "amd64"}); !errors.Is(err, ErrInvalidRelease) {
				t.Fatalf("error = %v, want ErrInvalidRelease", err)
			}
		})
	}
}

func TestCheckIgnoresGitHubSourceArchiveFields(t *testing.T) {
	t.Parallel()
	asset := "ptrack_1.2.4_linux_amd64.tar.gz"
	body := strings.Replace(releaseJSON("v1.2.4", false, false, asset), `"assets":`, `"tarball_url":"https://api.github.com/repos/ro-ag/ptrack/tarball/v1.2.4","zipball_url":"https://api.github.com/repos/ro-ag/ptrack/zipball/v1.2.4","assets":`, 1)
	client := fixtureClient(t, body)
	candidate, err := client.Check(context.Background(), "1.2.3", Target{GOOS: "linux", GOARCH: "amd64"})
	if err != nil {
		t.Fatal(err)
	}
	if candidate.Package.Name != asset || strings.Contains(candidate.Package.DownloadURL, "tarball") {
		t.Fatalf("source archive selected: %#v", candidate.Package)
	}
}

func TestCheckBoundsAndCancelsMetadataRequest(t *testing.T) {
	t.Parallel()
	t.Run("oversized", func(t *testing.T) {
		t.Parallel()
		client := fixtureClient(t, strings.Repeat("x", maxMetadataBytes+1))
		if _, err := client.Check(context.Background(), "1.2.3", Target{GOOS: "linux", GOARCH: "amd64"}); !errors.Is(err, ErrInvalidRelease) {
			t.Fatalf("error = %v, want ErrInvalidRelease", err)
		}
	})
	t.Run("canceled", func(t *testing.T) {
		t.Parallel()
		started := make(chan struct{})
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			close(started)
			<-r.Context().Done()
		}))
		defer server.Close()
		client := &Client{endpoint: server.URL, client: server.Client()}
		ctx, cancel := context.WithCancel(context.Background())
		done := make(chan error, 1)
		go func() {
			_, err := client.Check(ctx, "1.2.3", Target{GOOS: "linux", GOARCH: "amd64"})
			done <- err
		}()
		<-started
		cancel()
		if err := <-done; !errors.Is(err, context.Canceled) {
			t.Fatalf("error = %v, want context.Canceled", err)
		}
	})
	t.Run("client timeout", func(t *testing.T) {
		t.Parallel()
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			<-r.Context().Done()
		}))
		defer server.Close()
		client := &Client{endpoint: server.URL, client: &http.Client{Timeout: 20 * time.Millisecond}}
		if _, err := client.Check(context.Background(), "1.2.3", Target{GOOS: "linux", GOARCH: "amd64"}); !errors.Is(err, context.DeadlineExceeded) {
			t.Fatalf("error = %v, want deadline exceeded", err)
		}
	})
}

func TestCheckSendsBoundedGitHubRequestHeaders(t *testing.T) {
	t.Parallel()
	asset := "ptrack_1.2.4_linux_amd64.tar.gz"
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet || r.Header.Get("User-Agent") != userAgent ||
			r.Header.Get("Accept") != "application/vnd.github+json" ||
			r.Header.Get("X-GitHub-Api-Version") != "2022-11-28" {
			t.Errorf("unexpected request: method=%s headers=%v", r.Method, r.Header)
		}
		fmt.Fprint(w, releaseJSON("v1.2.4", false, false, asset))
	}))
	defer server.Close()
	client := &Client{endpoint: server.URL, client: server.Client()}
	if _, err := client.Check(context.Background(), "1.2.3", Target{GOOS: "linux", GOARCH: "amd64"}); err != nil {
		t.Fatal(err)
	}
}

func TestCheckRejectsMetadataRedirectAndStatus(t *testing.T) {
	t.Parallel()
	var destinationRequests atomic.Int32
	final := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		destinationRequests.Add(1)
		fmt.Fprint(w, `{}`)
	}))
	defer final.Close()
	redirect := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		http.Redirect(w, r, final.URL, http.StatusFound)
	}))
	defer redirect.Close()
	client := &Client{endpoint: redirect.URL, client: redirect.Client()}
	if _, err := client.Check(context.Background(), "1.2.3", Target{GOOS: "linux", GOARCH: "amd64"}); !errors.Is(err, ErrInvalidRelease) {
		t.Fatalf("redirect error = %v, want ErrInvalidRelease", err)
	}
	if got := destinationRequests.Load(); got != 0 {
		t.Fatalf("redirect destination received %d requests, want 0", got)
	}

	status := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusTooManyRequests)
	}))
	defer status.Close()
	client = &Client{endpoint: status.URL, client: status.Client()}
	if _, err := client.Check(context.Background(), "1.2.3", Target{GOOS: "linux", GOARCH: "amd64"}); err == nil || errors.Is(err, ErrInvalidRelease) {
		t.Fatalf("status error = %v", err)
	}
}

func fixtureClient(t *testing.T, body string) *Client {
	t.Helper()
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprint(w, body)
	}))
	t.Cleanup(server.Close)
	return &Client{endpoint: server.URL, client: server.Client()}
}

func releaseJSON(tag string, draft, prerelease bool, asset string) string {
	return fmt.Sprintf(`{
  "tag_name":%q,
  "body":"Safe release notes",
  "draft":%t,
  "prerelease":%t,
  "published_at":"2026-08-11T00:00:00Z",
  "assets":[
    {"name":%q,"browser_download_url":%q,"size":1024,"state":"uploaded"},
    {"name":"checksums.txt","browser_download_url":%q,"size":128,"state":"uploaded"}
  ]
}`,
		tag,
		draft,
		prerelease,
		asset,
		"https://github.com/ro-ag/ptrack/releases/download/"+tag+"/"+asset,
		"https://github.com/ro-ag/ptrack/releases/download/"+tag+"/checksums.txt",
	)
}

func duplicateAsset(body, name string) string {
	needle := fmt.Sprintf(`{"name":%q`, name)
	start := strings.Index(body, needle)
	if start < 0 {
		return body
	}
	end := strings.Index(body[start:], "}")
	if end < 0 {
		return body
	}
	asset := body[start : start+end+1]
	return strings.Replace(body, `"assets":[`, `"assets":[`+asset+`,`, 1)
}

func TestParseVersionStrictness(t *testing.T) {
	t.Parallel()
	for _, valid := range []string{"0.0.0", "1.2.3", "v1.2.3", "18446744073709551615.0.1"} {
		if _, err := parseVersion(valid, true); err != nil {
			t.Errorf("parseVersion(%q) = %v", valid, err)
		}
	}
	for _, invalid := range []string{"v", "1", "1.2", "1.2.3.4", "1.02.3", "-1.2.3", "1.2.3+meta", "1.2.3-rc.1", " 1.2.3", "1.2.3 ", "18446744073709551616.0.0"} {
		if _, err := parseVersion(invalid, true); err == nil {
			t.Errorf("parseVersion(%q) unexpectedly succeeded", invalid)
		}
	}
}

func TestCompareVersions(t *testing.T) {
	t.Parallel()
	for _, test := range []struct {
		left, right string
		want        int
	}{
		{left: "v1.2.4", right: "1.2.3", want: 1},
		{left: "1.2.3", right: "v1.2.3", want: 0},
		{left: "1.2.2", right: "1.2.3", want: -1},
	} {
		got, err := CompareVersions(test.left, test.right)
		if err != nil || got != test.want {
			t.Errorf("CompareVersions(%q, %q) = %d, %v; want %d", test.left, test.right, got, err, test.want)
		}
	}
	if _, err := CompareVersions("dev", "1.2.3"); err == nil {
		t.Fatal("CompareVersions accepted dev")
	}
}

func TestNewClientHasProductionSafetyDefaults(t *testing.T) {
	t.Parallel()
	client := NewClient()
	if client.endpoint != latestReleaseURL || client.client.Timeout != 15*time.Second || client.client.CheckRedirect == nil {
		t.Fatalf("unsafe production client: %#v", client)
	}
}

File: crates/ptrack-updater/src/discovery_test.rs
use std::sync::Arc;

use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpListener;
use tokio_util::sync::CancellationToken;

use super::discovery::{
    Client, Target, UpdateError, compare_versions, package_name, parse_version, select_candidate,
};

#[test]
fn strict_versions_and_exact_target_names_are_frozen() {
    assert_eq!(parse_version("v1.2.3", true).unwrap().to_string(), "1.2.3");
    for value in ["1.2", "1.2.3.4", "1.02.3", "1.2.3-beta", " 1.2.3", ""] {
        assert!(parse_version(value, true).is_err(), "accepted {value:?}");
    }
    assert!(compare_versions("1.2.3", "1.3.0").unwrap().is_lt());
    assert_eq!(
        package_name(
            &Target {
                os: "windows".to_owned(),
                arch: "arm64".to_owned(),
            },
            "1.2.3",
        )
        .unwrap(),
        "ptrack_1.2.3_windows_arm64.zip"
    );
    assert!(
        package_name(
            &Target {
                os: "freebsd".to_owned(),
                arch: "amd64".to_owned(),
            },
            "1.2.3"
        )
        .is_err()
    );
}

#[test]
fn release_notes_accept_exact_thirty_two_kibibyte_boundary_only() {
    let target = Target {
        os: "linux".to_owned(),
        arch: "amd64".to_owned(),
    };
    let current = parse_version("1.2.3", true).unwrap();
    assert!(select_candidate(&release_with_notes(32 << 10, &target), current, &target).is_ok());
    assert_eq!(
        select_candidate(
            &release_with_notes((32 << 10) + 1, &target),
            current,
            &target,
        )
        .unwrap_err(),
        UpdateError::InvalidRelease
    );
}

#[test]
fn release_publication_rejects_go_zero_time_but_accepts_unix_epoch() {
    let target = Target {
        os: "linux".to_owned(),
        arch: "amd64".to_owned(),
    };
    let current = parse_version("1.2.3", true).unwrap();
    assert_eq!(
        select_candidate(
            &release_with_publication("0001-01-01T00:00:00Z", &target),
            current,
            &target,
        )
        .unwrap_err(),
        UpdateError::InvalidRelease
    );
    assert!(
        select_candidate(
            &release_with_publication("1970-01-01T00:00:00Z", &target),
            current,
            &target,
        )
        .is_ok()
    );
}

#[tokio::test(flavor = "current_thread")]
async fn production_check_executes_fixed_headers_and_selects_exact_assets() {
    let target = Target {
        os: "linux".to_owned(),
        arch: "amd64".to_owned(),
    };
    let package = package_name(&target, "1.2.4").unwrap();
    let release = serde_json::json!({
        "tag_name": "v1.2.4",
        "body": "notes",
        "draft": false,
        "prerelease": false,
        "published_at": "2026-01-02T03:04:05Z",
        "assets": [
            {"name": package, "browser_download_url": format!("https://github.com/ro-ag/ptrack/releases/download/v1.2.4/{package}"), "size": 123, "state": "uploaded"},
            {"name": "checksums.txt", "browser_download_url": "https://github.com/ro-ag/ptrack/releases/download/v1.2.4/checksums.txt", "size": 80, "state": "uploaded"}
        ]
    })
    .to_string();
    let (endpoint, request) = serve_once(release.into_bytes()).await;
    let candidate = Client::with_endpoint(endpoint)
        .unwrap()
        .check(&CancellationToken::new(), "1.2.3", &target)
        .await
        .unwrap();
    assert_eq!(candidate.version, "1.2.4");
    assert_eq!(candidate.package.name, package);
    let request = request.await.unwrap().to_ascii_lowercase();
    assert!(request.contains("accept: application/vnd.github+json\r\n"));
    assert!(request.contains("user-agent: p-track-updater\r\n"));
    assert!(request.contains("x-github-api-version: 2022-11-28\r\n"));
}

#[tokio::test(flavor = "current_thread")]
async fn discovery_rejects_metadata_over_one_mebibyte() {
    let (endpoint, _) = serve_once(vec![b' '; (1 << 20) + 1]).await;
    let error = Client::with_endpoint(endpoint)
        .unwrap()
        .check(
            &CancellationToken::new(),
            "1.2.3",
            &Target {
                os: "linux".to_owned(),
                arch: "amd64".to_owned(),
            },
        )
        .await
        .unwrap_err();
    assert_eq!(error, UpdateError::InvalidRelease);
}

async fn serve_once(body: Vec<u8>) -> (String, tokio::task::JoinHandle<String>) {
    let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
    let address = listener.local_addr().unwrap();
    let body = Arc::new(body);
    let task = tokio::spawn(async move {
        let (mut stream, _) = listener.accept().await.unwrap();
        let mut request = Vec::new();
        let mut chunk = [0_u8; 1024];
        loop {
            let count = stream.read(&mut chunk).await.unwrap();
            request.extend_from_slice(&chunk[..count]);
            if count == 0 || request.windows(4).any(|window| window == b"\r\n\r\n") {
                break;
            }
        }
        let header = format!(
            "HTTP/1.1 200 OK\r\nContent-Length: {}\r\nConnection: close\r\n\r\n",
            body.len()
        );
        stream.write_all(header.as_bytes()).await.unwrap();
        stream.write_all(&body).await.unwrap();
        String::from_utf8(request).unwrap()
    });
    (format!("http://{address}/latest"), task)
}

fn release_with_notes(length: usize, target: &Target) -> Vec<u8> {
    release_fixture(&"x".repeat(length), "2026-01-02T03:04:05Z", target)
}

fn release_with_publication(publication: &str, target: &Target) -> Vec<u8> {
    release_fixture("", publication, target)
}

fn release_fixture(body: &str, publication: &str, target: &Target) -> Vec<u8> {
    let package = package_name(target, "1.2.4").unwrap();
    serde_json::to_vec(&serde_json::json!({
        "tag_name": "v1.2.4",
        "body": body,
        "draft": false,
        "prerelease": false,
        "published_at": publication,
        "assets": [
            {"name": package, "browser_download_url": format!("https://github.com/ro-ag/ptrack/releases/download/v1.2.4/{package}"), "size": 123, "state": "uploaded"},
            {"name": "checksums.txt", "browser_download_url": "https://github.com/ro-ag/ptrack/releases/download/v1.2.4/checksums.txt", "size": 80, "state": "uploaded"}
        ]
    }))
    .unwrap()
}
